### ISSUE TITLE
Mismatch of history vs optimizer reported optimal values

### ISSUE THREAD

**Issue by yannikschaelte** [ICB-DCM/pyPESTO#327]
When executing the rosenbrock.ipynb notebook on develop branch, I get massive complaints:

```
Function values from history and optimizer do not match: 1.3168387678656086, 2.0179911928533514
Parameters obtained from history and optimizer do not match: [0.99553374 0.98995083 0.97544456 0.94768476 0.89550744 0.8019584
 0.64185456 0.40180156 0.14794723 0.00844711], [0.9940831  0.98497745 0.96635783 0.93780681 0.87380595 0.74153588
 0.53395112 0.25523609 0.03957006 0.00100739]
Function values from history and optimizer do not match: 1.6349167203078765, 2.336760892779954
```

I guess something must still be wrong about history values vs. values returned by optimizers. Happens for various optimizers.

**Comment by LeonardSchmiester** [comment]
Had the same issues for more or less every benchmark model

**Comment by paulstapor** [comment]
Same here...

**Comment by yannikschaelte** [comment]
@LeonardSchmiester @paulstapor since when roughly did you observe this?

**Comment by paulstapor** [comment]
pulled and tested this morning... So, since today... But might have been introduced earlier...

**Comment by LeonardSchmiester** [comment]
I saw it for the benchmark optimization runs in mid-april with pyPESTO 0.0.12 I think.

**Comment by yannikschaelte** [comment]
did it happen for all optimizers, or only for specific ones (e.g. least squares)? (I think also for non-ls --)

**Comment by paulstapor** [comment]
Happened to me with TNC, Zheng model, optimization for pyPESTO PEtab...

**Comment by LeonardSchmiester** [comment]
I don't remember for sure, but I think it was multiple optimizers and more or less all models

**Comment by yannikschaelte** [comment]
oki. 
* I cannot find any problems in the history and optimizer files. 
* the warning may appear since mid-april since roughly then we started raising a warning at all in case the history contained a better value than the optimizer reported.
* I reran the problem, and I can only reproduce the error with optimizer TNC. Checking the objfun at those parameters indicates that the value reported in the history is **actually better**.
* No boundary conditions are hit.

![image](https://user-images.githubusercontent.com/31767307/81811330-c124d800-9524-11ea-956c-acbe38d37e6c.png)

**Comment by yannikschaelte** [comment]
So, currently it seems like TNC does not report the optimal value. however only sometimes: Out of 20 starts, the message appears only 1-2 times usually, more often for dim >>.  unless there is something strange going on which I have not found yet.

**Comment by yannikschaelte** [comment]
Update: Tested all scipy optimizers, it seems to happen for TNC and L-BFGS-B.

**Comment by yannikschaelte** [comment]
Also tested with a single start, problem remains. I.e. there can be no interference of values reported for different starts.

**Comment by paulstapor** [comment]
Okay, so TNC simply reports the last value, not the best found one, right?
Then... it might either make sense 
 * to add a postprocessing step after optimization
 * open a feature request / bug report for SciPy
 * ?

**Comment by FFroehlich** [comment]
What does the corresponding text output look like? I would expect this is a bug in TNC. The code is pretty old and I find it extremely difficult to follow whats going on.

The postprocessing is already in place when using history, thats why we found this issue in the first place.

**Comment by yannikschaelte** [comment]
@paulstapor @FFroehlich yes, TNC seems to report the last value (the optimal one is a few steps before), so I guess this is a bug in TNC. The code is quite comprehensive, but hard to follow, potentially the goto logic can mess things up.

In pyPESTO we are fine since we have the post-processing in place. The warning I would leave in place as a reminder that currently the algorithm is bugged.

**Comment by LeonardSchmiester** [comment]
I just experienced the same issue with the `powell` optimizer. Returns substantially larger values than the optimal along the history (`2.146` reported vs  `0.757` as best value). It also doesn't return the last value.

> In pyPESTO we are fine

At least for Powell, the suboptimal values are also returned by pyPESTO.

**Comment by FFroehlich** [comment]
> I just experienced the same issue with the `powell` optimizer. Returns substantially larger values than the optimal along the history (`2.146` reported vs `0.757` as best value). It also doesn't return the last value.
> 
> > In pyPESTO we are fine
> 
> At least for Powell, the suboptimal values are also returned by pyPESTO.

Is there a difference in objective function value when evaluating with or without sensitivities?

**Comment by yannikschaelte** [comment]
> In pyPESTO we are fine

Rigiht, when the optimizer reports a value, we currently use that value and ignore the history (raising a warning if there is a mismatch). It would be a policy/options decision whether we want to override from history in such cases I guess ... both would make sense I think ... maybe the Powell optimizer is simply bad ...

**Comment by LeonardSchmiester** [comment]
> > I just experienced the same issue with the `powell` optimizer. Returns substantially larger values than the optimal along the history (`2.146` reported vs `0.757` as best value). It also doesn't return the last value.
> > > In pyPESTO we are fine
> > 
> > 
> > At least for Powell, the suboptimal values are also returned by pyPESTO.
> 
> Is there a difference in objective function value when evaluating with or without sensitivities?

No, obj function values are the same with / without sensitivities.

**Comment by yannikschaelte** [comment]
Summary suggestion (from discussion if I remember correctly):

* We want to have both. Especially in benchmarks, we want the optimizer result, but sometimes the history value may just be so much better.

Thus:

* We have an option `history_beats_optimizer_result=False` (or similar) prioritizing by default the optimizer
* We ensure history values are admissible (within bounds!)

Q:

* When to raise a warning about mismatch? Always?

**Comment by dweindl** [comment]
Agreed that it's good to have both available. My guess would be that users mostly care about good parameters and less so about whether they were (what some optimizer considers) the final parameters, or whether they occurred elsewhere. So `history_beats_optimizer_result=True` would be a reasonable default for me. Can live with either, though.
Regarding warning: If we have a `history_beats_optimizer_result` option, I would not issue a warning anymore. If at all, then only debug-level logging.
